File: kakao.app/js/ked/view/menu/inputchoice.js
var _k_ = {extend: function (c,p) {for (var k in p) { if (Object.hasOwn(p, k)) c[k] = p[k] } function ctor() { this.constructor = c; } ctor.prototype = p.prototype; c.prototype = new ctor(); c.__super__ = p.prototype; return c;}, min: function () { var m = Infinity; for (var a of arguments) { if (Array.isArray(a)) {m = _k_.min.apply(_k_.min,[m].concat(a))} else {var n = parseFloat(a); if(!isNaN(n)){m = n < m ? n : m}}}; return m }, isStr: function (o) {return typeof o === 'string' || o instanceof String}, trim: function (s,c=' ') {return _k_.ltrim(_k_.rtrim(s,c),c)}, in: function (a,l) {return (typeof l === 'string' && typeof a === 'string' && a.length ? '' : []).indexOf.call(l,a) >= 0}, ltrim: function (s,c=' ') { while (_k_.in(s[0],c)) { s = s.slice(1) } return s}, rtrim: function (s,c=' ') {while (_k_.in(s.slice(-1)[0],c)) { s = s.slice(0, s.length - 1) } return s}}

var inputchoice

import kxk from "../../../kxk.js"
let post = kxk.post

import theme from "../../theme/theme.js"

import view from "../base/view.js"
import input from "../base/input.js"

import choices from "../menu/choices.js"


inputchoice = (function ()
{
    _k_.extend(inputchoice, view)
    function inputchoice (screen, name, features)
    {
        this.screen = screen
        this.name = name
    
        var _33_23_

        this["onWheel"] = this["onWheel"].bind(this)
        this["onMouse"] = this["onMouse"].bind(this)
        this["onKey"] = this["onKey"].bind(this)
        this["onChoicesAction"] = this["onChoicesAction"].bind(this)
        this["autoHide"] = this["autoHide"].bind(this)
        this["onInputAction"] = this["onInputAction"].bind(this)
        this["onInputChange"] = this["onInputChange"].bind(this)
        this["hide"] = this["hide"].bind(this)
        this["show"] = this["show"].bind(this)
        this["arrange"] = this["arrange"].bind(this)
        inputchoice.__super__.constructor.call(this,this.screen,this.name,features)
        this.autoHideInput = true
        this.isVisible = false
        this.isPopup = true
        this.input = new input(this.screen,`${this.name}_input`)
        this.choices = new choices(this.screen,`${this.name}_choices`,features)
        this.setColor('bg',theme.quicky.bg)
        this.setColor('frame',theme.quicky.frame)
        ;(this.choices.mapscr != null ? this.choices.mapscr.hide() : undefined)
        this.input.hide()
        this.choices.show()
        this.input.on('action',this.onInputAction)
        this.choices.on('action',this.onChoicesAction)
    }

    inputchoice.prototype["setColor"] = function (key, color)
    {
        inputchoice.__super__.setColor.call(this,key,color)
    
        if (key === 'bg')
        {
            this.input.setColor('bg',this.color.bg)
            this.input.setColor('empty',this.color.bg)
            return this.choices.setColor('bg',this.color.bg)
        }
    }

    inputchoice.prototype["inputIsActive"] = function ()
    {
        return this.input.hasFocus() || this.input.current().length
    }

    inputchoice.prototype["arrange"] = function ()
    {
        var cs, h, w, x, y

        x = parseInt(this.screen.cols / 4)
        y = parseInt(this.screen.rows / 4)
        w = parseInt(this.screen.cols / 2)
        h = parseInt(this.screen.rows / 2 - 4)
        cs = _k_.min(h,this.choices.numFiltered())
        this.input.layout(x + 2,y + 1,w - 4,1)
        this.choices.layout(x + 2,y + 3,w - 3,cs)
        return this.cells.layout(x,y,w,cs + 4)
    }

    inputchoice.prototype["show"] = function ()
    {
        inputchoice.__super__.show.call(this)
    
        if (this.choices.numFiltered())
        {
            this.choices.grabFocus()
        }
        else if (this.input.visible())
        {
            this.input.grabFocus()
        }
        return {redraw:true}
    }

    inputchoice.prototype["hide"] = function ()
    {
        var _89_23_

        ;(this.choices.mapscr != null ? this.choices.mapscr.hide() : undefined)
        return inputchoice.__super__.hide.call(this)
    }

    inputchoice.prototype["onInputChange"] = function (text)
    {
        if (text === this.choices.filterText)
        {
            return
        }
        this.choices.filter(text)
        this.choices.state.selectLine(0)
        this.choices.state.setMainCursor(0,0)
        this.choicesFiltered()
        return this.arrange()
    }

    inputchoice.prototype["onInputAction"] = function (action, text)
    {
        switch (action)
        {
            case 'right':
            case 'submit':
                return this.applyChoice(this.choices.current())

            case 'change':
                return this.onInputChange(text)

            case 'up':
            case 'down':
                return this.choices.moveSelection(action)

        }

    }

    inputchoice.prototype["autoHide"] = function ()
    {
        if (this.autoHideInput && !this.inputIsActive())
        {
            return this.input.hide()
        }
    }

    inputchoice.prototype["onChoicesAction"] = function (action, choice)
    {
        switch (action)
        {
            case 'click':
            case 'right':
            case 'space':
            case 'return':
                return this.applyChoice(this.choices.current())

            case 'hover':
                return this.autoHide()

        }

    }

    inputchoice.prototype["choicesFiltered"] = function ()
    {}

    inputchoice.prototype["currentChoice"] = function ()
    {
        var choice, _144_36_

        choice = ((_144_36_=this.choices.current()) != null ? _144_36_ : this.input.current())
        if (_k_.isStr(choice))
        {
            return choice = _k_.trim(choice)
        }
    }

    inputchoice.prototype["draw"] = function ()
    {
        if (this.hidden())
        {
            return
        }
        this.arrange()
        this.drawFrame()
        return this.drawChoices()
    }

    inputchoice.prototype["drawChoices"] = function ()
    {
        this.input.draw()
        return this.choices.draw()
    }

    inputchoice.prototype["drawFrame"] = function ()
    {
        var bg, fg

        fg = this.color.frame
        bg = this.color.bg
        if (this.input.visible())
        {
            return this.cells.draw_frame(0,0,-1,-1,{fg:fg,bg:bg,hdiv:[2]})
        }
        else
        {
            return this.cells.draw_frame(0,0,-1,-1,{fg:fg,bg:bg})
        }
    }

    inputchoice.prototype["moveFocus"] = function ()
    {
        if (this.choices.hasFocus())
        {
            this.input.grabFocus()
            return this.input.selectAll()
        }
        else
        {
            this.choices.grabFocus()
            return this.autoHide()
        }
    }

    inputchoice.prototype["onKey"] = function (key, event)
    {
        var result

        if (this.hidden())
        {
            return
        }
        switch (event.combo)
        {
            case 'tab':
                return this.moveFocus()

            case 'esc':
                post.emit('focus','editor')
                return this.hide()

        }

        if (result = this.choices.onKey(key,event))
        {
            if (event.char && !(_k_.in(event.char,' \n')))
            {
                this.input.grabFocus()
            }
            else
            {
                this.autoHide()
                return result
            }
        }
        if (result = this.input.onKey(key,event))
        {
            this.autoHide()
            return result
        }
        return true
    }

    inputchoice.prototype["onMouse"] = function (event)
    {
        var ret

        if (this.hidden())
        {
            return
        }
        ret = this.input.onMouse(event)
        if ((ret != null ? ret.redraw : undefined))
        {
            return ret
        }
        ret = this.choices.onMouse(event)
        if ((ret != null ? ret.redraw : undefined))
        {
            return ret
        }
        ret = inputchoice.__super__.onMouse.call(this,event)
        if ((ret != null ? ret.redraw : undefined))
        {
            return ret
        }
        if (event.type === 'press' && !this.hover)
        {
            post.emit('focus','editor')
            return this.hide()
        }
        return this.hover
    }

    inputchoice.prototype["onWheel"] = function (event)
    {
        var inside, ret

        if (this.hidden())
        {
            return
        }
        ret = this.input.onWheel(event)
        if ((ret != null ? ret.redraw : undefined))
        {
            return ret
        }
        ret = this.choices.onWheel(event)
        if ((ret != null ? ret.redraw : undefined))
        {
            return ret
        }
        inside = this.cells.isInsideEvent(event)
        if (inside)
        {
            return {redraw:false}
        }
    }

    return inputchoice
})()

export default inputchoice;

File: kakao.app/love/lua/view/menu/inputchoice.lua
--[[
    ███  ███   ███  ████████   ███   ███  █████████   ███████  ███   ███   ███████   ███   ███████  ████████  
    ███  ████  ███  ███   ███  ███   ███     ███     ███       ███   ███  ███   ███  ███  ███       ███       
    ███  ███ █ ███  ████████   ███   ███     ███     ███       █████████  ███   ███  ███  ███       ███████   
    ███  ███  ████  ███        ███   ███     ███     ███       ███   ███  ███   ███  ███  ███       ███       
    ███  ███   ███  ███         ███████      ███      ███████  ███   ███   ███████   ███   ███████  ████████  

    a combination of input and choices
    the input is usually hidden and only shown when user types text to filter the choices
--]]

view = require "view.base.view"
input = require "view.base.input"


local inputchoice = class("inputchoice", view)
    


function inputchoice:init(name, features) 
        view.init(self, name, features)
        
        self.autoHideInput = true
        self.isVisible = false
        self.isPopup = true
        
        self.input = input("" .. tostring(self.name) .. "_input")
        self.choices = choices("" .. tostring(self.name) .. "_choices", features)
        
        self:setColor('bg', theme.quicky.bg)
        self:setColor('frame', theme.quicky.frame)
        
        if self.choices.mapscr then 
            self.choices.mapscr:hide()
        end
        
        self.input:hide()
        self.choices:show()
        
        self.input:on('action', self.onInputAction, self)
        self.choices:on('action', self.onChoicesAction, self)
        return self
    end


function inputchoice:setColor(key, clr) 
        view.setColor(self, key, clr)
        
        if (key == 'bg') then 
            self.input:setColor('bg_blur', self.color.bg)
            return self.choices:setColor('bg', self.color.bg)
        end
    end


function inputchoice:inputIsActive() 
    return (self.input:hasFocus() or (#self.input:current() > 0))
    end

--  ███████   ████████   ████████    ███████   ███   ███   ███████   ████████
-- ███   ███  ███   ███  ███   ███  ███   ███  ████  ███  ███        ███     
-- █████████  ███████    ███████    █████████  ███ █ ███  ███  ████  ███████ 
-- ███   ███  ███   ███  ███   ███  ███   ███  ███  ████  ███   ███  ███     
-- ███   ███  ███   ███  ███   ███  ███   ███  ███   ███   ███████   ████████


function inputchoice:arrange() 
        local x = floor((self.screen.cols / 4))
        local y = floor((self.screen.rows / 4))
        local w = floor((self.screen.cols / 2))
        local h = floor(((self.screen.rows / 2) - 4))
        
        local cs = min(h, self.choices:numFiltered())
        
        self.input:layout((x + 2), (y + 1), (w - 4), 1)
        self.choices:layout((x + 2), (y + 3), (w - 3), cs)
        return self.cells:layout(x, y, w, (cs + 4))
    end

--  0000000  000   000   0000000   000   000
-- 000       000   000  000   000  000 0 000
-- 0000000   000000000  000   000  000000000
--      000  000   000  000   000  000   000
-- 0000000   000   000   0000000   00     00


function inputchoice:show() 
        view.show(self)
        
        if self.choices:numFiltered() then 
            self.choices:grabFocus()
        elseif self.input:visible() then 
            self.input:grabFocus()
        end
        
        return true
    end


function inputchoice:hide() 
        if self.choices.mapscr then 
            self.choices.mapscr:hide()
        end
        
        return view.hide(self)
    end

-- 000  000   000  00000000   000   000  000000000  
-- 000  0000  000  000   000  000   000     000     
-- 000  000 0 000  00000000   000   000     000     
-- 000  000  0000  000        000   000     000     
-- 000  000   000  000         0000000      000     


function inputchoice:onInputChange(text) 
        if (text == self.choices.filterText) then return end
        
        self.choices.filter(text)
        self.choices.state.selectLine(0)
        
        self.choices.state.setMainCursor(0, 0)
        self:choicesFiltered()
        return self:arrange()
    end


function inputchoice:onInputAction(action, text) 
        if (action == 'submit') then 
    return self:applyChoice(self.choices:current())
        elseif (action == 'change') then 
    return self:onInputChange(text)
        elseif (action == 'up') or (action == 'down') then 
    return self.choices:moveSelection(action)
        end
    end


function inputchoice:autoHide() 
    if (self.autoHideInput and not self:inputIsActive()) then 
    return self.input:hide()
                               end
    end

--  0000000  000   000   0000000   000   0000000  00000000   0000000  
-- 000       000   000  000   000  000  000       000       000       
-- 000       000000000  000   000  000  000       0000000   0000000   
-- 000       000   000  000   000  000  000       000            000  
--  0000000  000   000   0000000   000   0000000  00000000  0000000   


function inputchoice:onChoicesAction(action, choice) 
        if (action == 'click') or (action == 'right') or (action == 'space') or (action == 'return') then 
    return self:applyChoice(self.choices:current())
        elseif (action == 'hover') then 
    return self:autoHide()
        elseif (action == 'boundary') or (action == 'left') then 
                if self.input:visible() then 
                    self.input:grabFocus()
                    return true
                end
        end
    end


function inputchoice:choicesFiltered() 
    
    end


function inputchoice:currentChoice() 
        local choice = (self.choices:current() or self.input:current())
        choice = (function () 
    if is(choice, "string") then 
    return kstr.trim(choice)
                                   end
end)()
        return choice
    end

-- 0000000    00000000    0000000   000   000
-- 000   000  000   000  000   000  000 0 000
-- 000   000  0000000    000000000  000000000
-- 000   000  000   000  000   000  000   000
-- 0000000    000   000  000   000  00     00


function inputchoice:draw() 
        if self:hidden() then return end
        
        self:arrange()
        self:drawFrame()
        return self:drawChoices()
    end


function inputchoice:drawChoices() 
        self.input:draw(self.screen)
        return self.choices:draw(self.screen)
    end


function inputchoice:drawFrame() 
        local sy = 1
        
        if self.input:visible() then 
            local outer = (function () 
    if self.input:hasFocus() then 
    return color.brighten(self.color.bg) else 
    return self.color.bg
                    end
end)()
            local inner = (function () 
    if self.input:hasFocus() then 
    return self.input.color.bg_focus else 
    return self.input.color.bg_blur
                    end
end)()
            
            self.cells:draw_rounded_border(0, 0, -1, 2, {fg = outer})
            
            -- add left and right half-width padding to input
            self.cells:draw_vertical_padding(1, 1, outer, inner)
            self.cells:draw_vertical_padding((self.cells.cols - 2), 1, inner, outer)
            
            sy = 2
        end
        
        return self.cells:draw_rounded_border(1, sy, -1, -1, {fg = self.color.bg, zLayer = 1001})
    end


function inputchoice:moveFocus() 
        if self.choices:hasFocus() then 
            self.input:grabFocus()
            return self.input:selectAll()
        else 
            self.choices:grabFocus()
            return self:autoHide()
        end
    end

-- 000   000  00000000  000   000
-- 000  000   000        000 000
-- 0000000    0000000     00000
-- 000  000   000          000
-- 000   000  00000000     000


function inputchoice:onKey(key, event) 
        if self:hidden() then return end
        
        if (event.combo == 'tab') then 
                return self:moveFocus()
        elseif (event.combo == 'esc') then 
                post:emit('focus', 'editor')
                return self:hide()
        end
        
        if self.input:hasFocus() then 
            local result = self.input:onKey(key, event)
            if result then 
                self:autoHide()
                return result
            end
            
            print("focused input didn't handle key " .. tostring(key) .. "?")
        end
        
        local result = self.choices:onKey(key, event)
        
        if result then 
            if ((valid(event.char) and (event.char ~= ' ')) and (event.char ~= '\n')) then 
                self.input:grabFocus()
                result = self.input:onKey(key, event)
                if result then 
                    self:autoHide()
                    return result
                end
            else 
                if (key == 'down') then self.choices:grabFocus() end
                self:autoHide()
                return result
            end
        end
        
        return true
    end

-- 00     00   0000000   000   000   0000000  00000000  
-- 000   000  000   000  000   000  000       000       
-- 000000000  000   000  000   000  0000000   0000000   
-- 000 0 000  000   000  000   000       000  000       
-- 000   000   0000000    0000000   0000000   00000000  


function inputchoice:onMouse(event) 
        if self:hidden() then return end
        
        if self.input:onMouse(event) then return true end
        if self.choices:onMouse(event) then return true end
        if view.onMouse(self, event) then return true end
        
        if ((event.type == 'press') and not self.hover) then 
            post:emit('focus', 'editor')
            self:hide()
            return -- give other views a chance to handle the closing click
        end
        
        return self.hover
    end

-- 000   000  000   000  00000000  00000000  000      
-- 000 0 000  000   000  000       000       000      
-- 000000000  000000000  0000000   0000000   000      
-- 000   000  000   000  000       000       000      
-- 00     00  000   000  00000000  00000000  0000000  


function inputchoice:onWheel(event) 
        if self:hidden() then return end
        if self.input:onWheel(event) then return true end
        if self.choices:onWheel(event) then return true end
        
        local inside = self.cells:isInsideEvent(event)
        if inside then 
    return true
        end
    end

return inputchoice